### DISCUSSION TITLE
Process jobs based on ressource lock.

### DISCUSSION THREAD

**Discussion by clouedoc** [taskforcesh/bullmq#579] [Q&A]
Hello to the people reading this message.

I currently have a use case where jobs must access shared ressources, but only one job can access the given ressource at a time. Jobs are assigned to a ressource (they can't use another one)

I made some nice images so that it's easier to understand the situation

![image](https://user-images.githubusercontent.com/13921610/120720677-cad10300-c4cc-11eb-80b9-64696c02d686.png)
![image](https://user-images.githubusercontent.com/13921610/120720699-d45a6b00-c4cc-11eb-8ab3-67e757d3bec9.png)
![image](https://user-images.githubusercontent.com/13921610/120720711-d91f1f00-c4cc-11eb-9513-31bb93eb7ec7.png)

Let me know if it isn't clear enough.

**Solutions I've thought of**

- Processing all jobs and looking up a shared mutex for antennas. Jobs would get delayed by X minutes if antenna not available. It looks like a bad solution for a number of reasons.

**Dream solution**

The best would be to have some way of specifying a "filter" function when instantiating the worker.
E.g.:
```ts
new Worker("queue-name", (job) => console.log('received job: ' job), {
  filter: async (job): boolean => isAntennaAvailable(job.data.antennaId)
})
```

**Comment by clouedoc** [comment]
I've seen someone talk about `groupKey`: https://github.com/taskforcesh/bullmq/discussions/433#discussioncomment-818905

It looks like a nice solution, but I need to investigate it a bit to know if it's what I'm really looking for.

**Reply by manast** [reply]
In your filter proposal, say a job is filtered out, what should happen with that job after that?

**Reply by clouedoc** [reply]
Filtered out jobs shall be delayed to get tested again by the filter in a few minutes (interval should be set by user).

**Reply by manast** [reply]
@clouedoc in that case, wouldn't it solve the requirement if it was possible to move the job to the delay set inside the job processor instead of letting it complete?

**Reply by clouedoc** [reply]
@manast yeah, delaying the job by X would be the best (to keep nice statistics about job completions/failures). 

It would also be nice to have a "job delayed because of filter" event.

**Reply by manast** [reply]
Have you tried retries and custom backoffs? Maybe you can just throw from the job processor and then in the custom backoff determine how much to delay the job. If some feature is missing in custom backoffs I am willing to add it so that you can fullfil your requirements.

**Reply by clouedoc** [reply]
I did not try this, thanks for the hint. I'll tell you how it has gone from a developer experience point of view so that you can decide what to do for your project.
